### ISSUE TITLE
[dart:io] - stdin.readLineSync on Windows Powershell keeps waiting until something being typed

### ISSUE THREAD

**Issue by codesculpture** [dart-lang/sdk#54588]
```dart
void main() {
    final a = stdin.readLineSync();
    print(a);
}
```
If The above program  executed on windows powershell, and if u press enter it would just move the cursor to next line and keep repeatedly do this until u type something, this not letting to pass empty input. Originally Founded from mason_logger https://github.com/felangel/mason/issues/1189

**Comment by eseidel** [comment]
https://github.com/dart-lang/sdk/blob/main/sdk/lib/io/stdio.dart#L56 is the code in Dart, it looks like it tries to support Windows, maybe powershell acts different from cmd 🤷

**Comment by codesculpture** [comment]
> https://github.com/dart-lang/sdk/blob/main/sdk/lib/io/stdio.dart#L56 is the code in Dart, it looks like it tries to support Windows, maybe powershell acts different from cmd 🤷 

Yes powershell is exceptional for everything🥲

**Comment by eseidel** [comment]
Thanks for looking at this @codesculpture!  That's a pretty small amount of code, which should be possible to lift into a stand-alone Dart file and debug.  Let me know if you need any help getting set up to debug.  We're reachable on various Flutter related Discords including https://discord.gg/shorebird

**Comment by codesculpture** [comment]
Hmm, the `enter` being interpted as 10 in non-powershell and 13 in powershell which making this issue.

**Comment by codesculpture** [comment]
```cpp
#include<windows.h>
#include<stdint.h>
#include<iostream>

bool ReadByte(int* byte) {
  HANDLE h = GetStdHandle(STD_INPUT_HANDLE);
  uint8_t buffer[1];
  DWORD read = 0;
  BOOL success = ReadFile(h, buffer, 1, &read, nullptr);
  if (!success && (GetLastError() != ERROR_BROKEN_PIPE)) {
    return false;
  }
  *byte = (read == 1) ? buffer[0] : -1;
  return true;
}

int main() {
        int a;
        ReadByte(&a);
        std::cout << a;
}
```
After Executing this, pressing enter would in powershell would print 13 but 10 in git bash and command prompt(which is right?, i believe)

**Comment by codesculpture** [comment]
https://github.com/dart-lang/sdk/blob/main/sdk/lib/io/stdio.dart#L77-L87

I think this is what supposed to get executed but `crNewLine` is evaluated to false.

Is `crNewLine` to detect carriage return with new line specifcally for powershell?

**Comment by codesculpture** [comment]
https://github.com/dart-lang/sdk/blob/main/runtime/bin/stdio_win.cc#L78

`lineMode` needs to be `false` on poweshell?

**Comment by codesculpture** [comment]
> When a console is created, all input modes except ENABLE_WINDOW_INPUT and ENABLE_VIRTUAL_TERMINAL_INPUT are enabled by default.

From https://learn.microsoft.com/en-us/windows/console/getconsolemode 

This makes always ENABLE_LINE_INPUT and ENABLE_ECHO_INPUT to be true. But we both needed to be disabled. 

I think we don't need to bother about `lineMode` since i think not necessarily to check while doing `readLineSync`

**Comment by a-siva** [comment]
//cc @brianquinlan

**Comment by codesculpture** [comment]
Then name itself `readLineByte` referring to reading a single line not multi line, so we need to return the data when switched to new line regardless of `lineMode`

**Comment by codesculpture** [comment]
If `ENABLE_LINE_INPUT` is disabled then `ENABLE_ECHO_INPUT` must disabled, but when `ENABLE_ECHO_INPUT` is disabled, user cant see what they typed, usually it used for password entering purposes. So considering these, why we need to expect `lineMode == false` to determine `crIsNewLine`🤔

**Comment by mraleph** [comment]
I was answering @codesculpture question on Discord which made me a bit curious about this problem. I could not reproduce this problem on my Windows 10 box even with newest Powershell. Eventually I have discovered that it only reproduces with Windows Terminal application, which I did not have installed.

If you play with reading byte-by-byte from stdin in C under Windows Terminal you will immediately notice that it is utterly broken: sometimes you press enter and the reader gets CR LF, sometimes you press enter and the reader only gets CR and LF is mysteriously dropped. 

Which lead me to discover https://github.com/microsoft/terminal/issues/16223 - so this is a bug in Windows Terminal and not in Dart.
